cmd/podman/completion/completion.go

```diff
@@ -0,0 +1,368 @@
+
+	case containersExited != nil:
+		return containersCreated, cobra.ShellCompDirectiveNoFileComp
+
+	default:
+		return nil, cobra.ShellCompDirectiveError
+	}
+}
+
+//Autocomplete all pod names.
+func AutocompletePods(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return getPods("all")
+}
+
+//Autocomplete only created pod names.
+/*func AutocompletePodsCreated(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return getPods("created")
+}*/
+
+//Autocomplete only exited pod names.
+/*func AutocompletePodsExited(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return getPods("exited")
+}*/
+
+//Autocomplete only running pod names.
+func AutocompletePodsRunning(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return getPods("running")
+}
+
+//Autocomplete only dead pod names.
+/*func AutocompletePodsDead(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return getPods("dead")
+}*/
+
+//Autocomplete only created and exited pod names.
+/*func AutocompletePodsStartable(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	podsCreated, _ := getPods("created")
+	podsExited, _ := getPods("exited")
+
+	switch {
+	case podsCreated != nil && podsExited != nil:
+		//combine both
+		return append(podsCreated, podsExited...), cobra.ShellCompDirectiveNoFileComp
+
+	case podsCreated != nil:
+		return podsExited, cobra.ShellCompDirectiveNoFileComp
+
+	case podsExited != nil:
+		return podsCreated, cobra.ShellCompDirectiveNoFileComp
+
+	default:
+		return nil, cobra.ShellCompDirectiveError
+	}
+}*/
+
+//Autocomplete container names and pod names.
+func AutocompleteContainersAndPods(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	containers, _ := getContainers("all")
+	pods, _ := getPods("all")
+
+	switch {
+	case containers != nil && pods != nil:
+		//combine both
+		return append(containers, pods...), cobra.ShellCompDirectiveNoFileComp
+
+	case containers != nil:
+		return pods, cobra.ShellCompDirectiveNoFileComp
+
+	case pods != nil:
+		return containers, cobra.ShellCompDirectiveNoFileComp
+
+	default:
+		return nil, cobra.ShellCompDirectiveError
+	}
+}
+
+//Autocomplete container names and pod names.
+func AutocompleteContainersAndImages(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	containers, _ := getContainers("all")
+	images, _ := getImages(toComplete)
+
+	switch {
+	case containers != nil && images != nil:
+		//combine both
+		return append(containers, images...), cobra.ShellCompDirectiveNoFileComp
+
+	case containers != nil:
+		return images, cobra.ShellCompDirectiveNoFileComp
+
+	case images != nil:
+		return containers, cobra.ShellCompDirectiveNoFileComp
```

Comment:
I think the logic here is reversed

Reply:
Actually, thinking about it, what does `append()` do with nil? Wouldn't it be cleaner to just eliminate the `switch` entirely and trust that nil will be handled correctly?

Reply:
`append(nil, entry)` will produce a 1-length array containing `entry`. I haven't personally tested `append(entry, nil)` but I'm fairly certain it will produce the original array, unchanged.

Reply:
I tested this and it seems to work perfectly fine:
```diff
index f2b787b5b..91336cdba 100644
--- a/cmd/podman/completion/completion.go
+++ b/cmd/podman/completion/completion.go
@@ -309,20 +309,7 @@ func AutocompleteContainersAndImages(cmd *cobra.Command, args []string, toComple
        containers, _ := getContainers("all")
        images, _ := getImages(toComplete)

-       switch {
-       case containers != nil && images != nil:
-               //combine both
-               return append(containers, images...), cobra.ShellCompDirectiveNoFileComp
-
-       case containers != nil:
-               return images, cobra.ShellCompDirectiveNoFileComp
-
-       case images != nil:
-               return containers, cobra.ShellCompDirectiveNoFileComp
-
-       default:
-               return nil, cobra.ShellCompDirectiveError
-       }
+       return append(containers, images...), cobra.ShellCompDirectiveNoFileComp
 }

 //Autocomplete volumes.
```
Tested with `podman inspect` and:
* no images, no containers
* 1 image, no containers
* 2 images, no containers
* 2 images, 1 container
* 2 images, 2 containers

Did not test with 0 images, 2 containers :-)

I think the above is much more maintainable; it could also be applied to several other places in the code.

Reply:
yeah thats is mush better